File: Java/백준/Gold/17387. 선분 교차 2/선분 교차 2.java
import java.io.*;
import java.util.StringTokenizer;

public class Main {

    static class Point {
        long x, y;

        Point(long x, long y) {
            this.x = x;
            this.y = y;
        }
    }

    static int ccw(Point a, Point b, Point c) {
        long op = (b.x - a.x) * (c.y - a.y) - (b.y - a.y) * (c.x - a.x);
        if (op > 0) return 1;
        if (op < 0) return -1;
        return 0;
    }

    static boolean isOverlapping(Point a, Point b, Point c, Point d) {
        return Math.min(a.x, b.x) <= Math.max(c.x, d.x) &&
               Math.max(a.x, b.x) >= Math.min(c.x, d.x) &&
               Math.min(a.y, b.y) <= Math.max(c.y, d.y) &&
               Math.max(a.y, b.y) >= Math.min(c.y, d.y);
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        long x1 = Long.parseLong(st.nextToken());
        long y1 = Long.parseLong(st.nextToken());
        long x2 = Long.parseLong(st.nextToken());
        long y2 = Long.parseLong(st.nextToken());

        st = new StringTokenizer(br.readLine());
        long x3 = Long.parseLong(st.nextToken());
        long y3 = Long.parseLong(st.nextToken());
        long x4 = Long.parseLong(st.nextToken());
        long y4 = Long.parseLong(st.nextToken());

        Point A = new Point(x1, y1);
        Point B = new Point(x2, y2);
        Point C = new Point(x3, y3);
        Point D = new Point(x4, y4);

        int ab_c = ccw(A, B, C);
        int ab_d = ccw(A, B, D);
        int cd_a = ccw(C, D, A);
        int cd_b = ccw(C, D, B);

        if (ab_c * ab_d == 0 && cd_a * cd_b == 0) {
            if (isOverlapping(A, B, C, D)) {
                System.out.println(1);
            } else {
                System.out.println(0);
            }
        } else {
            if (ab_c * ab_d <= 0 && cd_a * cd_b <= 0) {
                System.out.println(1);
            } else {
                System.out.println(0);
            }
        }
    }
}

File: C11/백준/Gold/17387. 선분 교차 2/선분 교차 2.c
#include <stdio.h>

typedef struct {
    long long x, y;
} Point;

int ccw(Point a, Point b, Point c) {
    long long op = (b.x - a.x) * (c.y - a.y) - (b.y - a.y) * (c.x - a.x);
    if (op > 0) return 1;
    if (op < 0) return -1;
    return 0;
}

int in_range(long long a, long long b, long long c) {
    if (a > b) {
        long long tmp = a;
        a = b;
        b = tmp;
    }
    return a <= c && c <= b;
}

int isOverlapping(Point a, Point b, Point c, Point d) {
    if (a.x > b.x || (a.x == b.x && a.y > b.y)) {
        Point tmp = a; a = b; b = tmp;
    }
    if (c.x > d.x || (c.x == d.x && c.y > d.y)) {
        Point tmp = c; c = d; d = tmp;
    }
    return !(b.x < c.x || d.x < a.x || b.y < c.y || d.y < a.y);
}

int main() {
    Point a, b, c, d;
    scanf("%lld %lld %lld %lld", &a.x, &a.y, &b.x, &b.y);
    scanf("%lld %lld %lld %lld", &c.x, &c.y, &d.x, &d.y);

    int ab_c = ccw(a, b, c);
    int ab_d = ccw(a, b, d);
    int cd_a = ccw(c, d, a);
    int cd_b = ccw(c, d, b);

    if (ab_c * ab_d == 0 && cd_a * cd_b == 0) {
        if (
            in_range(a.x, b.x, c.x) || in_range(a.x, b.x, d.x) ||
            in_range(c.x, d.x, a.x) || in_range(c.x, d.x, b.x)
        ) {
            if (
                in_range(a.y, b.y, c.y) || in_range(a.y, b.y, d.y) ||
                in_range(c.y, d.y, a.y) || in_range(c.y, d.y, b.y)
            ) {
                printf("1\n");
                return 0;
            }
        }
        printf("0\n");
    } else {
        if (ab_c * ab_d <= 0 && cd_a * cd_b <= 0)
            printf("1\n");
        else
            printf("0\n");
    }

    return 0;
}